### DISCUSSION TITLE
[QUESTION] How to convert a checkpoint to virtual pipeline format checkpoint

### DISCUSSION THREAD

**Discussion by ZacWang** [NVIDIA/Megatron-LM#1249] [Q&A]
**Your question**
I'm using `tools/checkpoint/convert.py` to convert a llama model to mcore model format for training. The `tools/checkpoint/loader_mcore.py` support virtual pipeline model loading, but `tools/checkpoint/saver_mcore.py` doesn't support to save a virtual pipeline model.

Do I have any other way to do this convert? Or do I need to modify `saver_mcore.py` to support this? Maybe with support for args like `target_num_layers_per_virtual_pipeline_stage` and `target_virtual_pipeline_model_parallel_size`?

**Comment by Advait251206** [comment]
As far as I'm aware, `tools/checkpoint/convert.py` doesn't currently support converting directly into a checkpoint that uses **virtual pipeline model parallelism (VPP)**.

While `loader_mcore.py` can load checkpoints with virtual pipeline support, `saver_mcore.py` only writes the standard pipeline-parallel checkpoint layout and doesn't expose options such as:

- `target_virtual_pipeline_model_parallel_size`
- `target_num_layers_per_virtual_pipeline_stage`

So, unless support has been added in a newer revision, there isn't an official conversion path for this.

Your options are essentially:

1. **Modify `saver_mcore.py`** to partition layers according to the desired virtual pipeline configuration and emit the checkpoint metadata expected by VPP. This is likely the most direct approach if you need an offline conversion tool.

2. **Convert to a standard MCore checkpoint first**, then launch training with the desired virtual pipeline settings and save a new checkpoint. Depending on your training setup and Megatron-LM version, the framework may rewrite the checkpoint in the appropriate format after initialization, though this isn't a guaranteed workflow for every configuration.

If your goal is a one-step offline conversion from a Hugging Face/LLaMA checkpoint directly into a VPP-enabled MCore checkpoint, I don't believe that functionality currently exists in the official conversion tools.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
